bitcoin/src/bip32.rs

```diff
@@ -1007,29 +1355,41 @@ mod tests {
     ) {
         let mut sk = Xpriv::new_master(network, seed).unwrap();
         let mut pk = Xpub::from_priv(secp, &sk);
-
         // Check derivation convenience method for Xpriv
         assert_eq!(&sk.derive_priv(secp, &path).to_string()[..], expected_sk);
 
+        // Tries to convert the path into a normal path, if it fails, it should return an error
+        match path.to_normal() {
+            Ok(path) => {
+                assert_eq!(&pk.derive_pub(secp, &path).to_string()[..], expected_pk);
+            }
+            Err(e) => {
+                assert_eq!(e, Error::CannotDeriveFromHardenedKey);
+            }
+        }
+
         // Check derivation convenience method for Xpub, should error
         // appropriately if any ChildNumber is hardened
-        if path.0.iter().any(|cnum| cnum.is_hardened()) {
-            assert_eq!(pk.derive_pub(secp, &path), Err(Error::CannotDeriveFromHardenedKey));
-        } else {
-            assert_eq!(&pk.derive_pub(secp, &path).unwrap().to_string()[..], expected_pk);
-        }
+        //if path.0.iter().any(|cnum| cnum.is_hardened()) {
+        //    assert_eq!(pk.try_derive_pub(secp, &path), Err(Error::CannotDeriveFromHardenedKey));
+        //} else {
+        //    assert_eq!(&pk.derive_pub(secp, path.into()).to_string()[..], expected_pk);
+        //}
 
         // Derive keys, checking hardened and non-hardened derivation one-by-one
-        for &num in path.0.iter() {
+        for num in path.into_iter() {
             sk = sk.ckd_priv(secp, num);
             match num {
                 Normal { .. } => {
-                    let pk2 = pk.ckd_pub(secp, num).unwrap();
+                    let pk2 = pk.internal_derive_pub(secp, num).unwrap();
```

Comment:
Tests should not access internal functions.

Reply:
Again, thats kinda out of scope. lets work with the tests on another PR.

Reply:
Why not call `derive_pub` then?

Reply:
Because we already have a `derive pub` they have different purposes.

Reply:
> Again, thats kinda out of scope. lets work with the tests on another PR. I agree with you btw.

I think we should add tests here. What about using underscores? `_derive_pub`?

Reply:
@jaoleal is there already a test that tests `derive_pub`? I haven't noticed.